File: solutions/java/src/main/java/com/epi/ValidIPAddress.java
package com.epi;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ValidIPAddress {
  // @include
  public static List<String> getValidIpAddress(String s) {
    List<String> result = new ArrayList<>();
    for (int i = 1; i < 4 && i < s.length(); ++i) {
      String first = s.substring(0, i);
      if (isValidPart(first)) {
        for (int j = 1; i + j < s.length() && j < 4; ++j) {
          String second = s.substring(i, i + j);
          if (isValidPart(second)) {
            for (int k = 1; i + j + k < s.length() && k < 4; ++k) {
              String third = s.substring(i + j, i + j + k);
              String fourth = s.substring(i + j + k);
              if (isValidPart(third) && isValidPart(fourth)) {
                result.add(first + "." + second + "." + third + "." + fourth);
              }
            }
          }
        }
      }
    }
    return result;
  }

  private static boolean isValidPart(String s) {
    if (s.length() > 3) {
      return false;
    }
    // "00", "000", "01", etc. are not valid, but "0" is valid.
    if (s.startsWith("0") && s.length() > 1) {
      return false;
    }
    int val = Integer.parseInt(s);
    return val <= 255 && val >= 0;
  }
  // @exclude

  public static void main(String[] args) {
    if (args.length == 1) {
      List<String> result = getValidIpAddress(args[0]);
      System.out.println(result);
    }
    List<String> res1 = getValidIpAddress("255255255255");
    System.out.println(res1);
    assert(res1.size() == 1);
    assert(res1.get(0).equals("255.255.255.255"));
    List<String> res2 = getValidIpAddress("19216811");
    System.out.println(res2);
    assert(res2.size() == 9);
    List<String> res3 = getValidIpAddress("1111");
    System.out.println(res3);
    assert(res3.size() == 1);
    assert(res3.get(0).equals("1.1.1.1"));
    List<String> res4 = getValidIpAddress("11000");
    System.out.println(res4);
    Collections.sort(res4);
    assert(res4.size() == 2);
    assert(res4.get(0).equals("1.10.0.0"));
    assert(res4.get(1).equals("11.0.0.0"));
  }
}

File: solutions/cpp/valid-IP-address.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <string>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::stoi;
using std::string;
using std::uniform_int_distribution;
using std::vector;

bool IsValidPart(const string& s);

// @include
vector<string> GetValidIPAddress(const string& s) {
  vector<string> result;
  for (size_t i = 1; i < 4 && i < s.size(); ++i) {
    auto first = s.substr(0, i);
    if (IsValidPart(first)) {
      for (size_t j = 1; i + j < s.size() && j < 4; ++j) {
        auto second = s.substr(i , j);
        if (IsValidPart(second)) {
          for (size_t k = 1; i + j + k < s.size() && k < 4; ++k) {
            auto third = s.substr(i + j, k), fourth = s.substr(i + j + k);
            if (IsValidPart(third) && IsValidPart(fourth)) {
              result.emplace_back(first + "." + second + "." + third + "." +
                                  fourth);
            }
          }
        }
      }
    }
  }
  return result;
}

bool IsValidPart(const string& s) {
  if (s.size() > 3) {
    return false;
  }
  // "00", "000", "01", etc. are not valid, but "0" is valid.
  if (s.front() == '0' && s.size() > 1) {
    return false;
  }
  int val = stoi(s);
  return val <= 255 && val >= 0;
}
// @exclude

int main(int argc, char** argv) {
  if (argc == 2) {
    auto result = GetValidIPAddress(argv[1]);
    for (const auto& s : result) {
      cout << s << endl;
    }
  }
  auto res1 = GetValidIPAddress("255255255255");
  for (const auto& s : res1) {
    cout << s << endl;
  }
  auto res2 = GetValidIPAddress("19216811");
  for (const auto& s : res2) {
    cout << s << endl;
  }
  return 0;
}
